src/main/catalog-sources/kubeconfig-sync.ts

```diff
@@ -0,0 +1,249 @@
+    }
+  }
+
+  @action
+  protected stopOldSync(filePath: string): void {
+    if (!this.sources.delete(filePath)) {
+      // already stopped
+      return void logger.debug(`${logPrefix} no syncing file/folder to stop`, { filePath });
+    }
+
+    logger.info(`${logPrefix} stopping sync of file/folder`, { filePath });
+    logger.debug(`${logPrefix} ${this.sources.size} files/folders watched`, { files: Array.from(this.sources.keys()) });
+  }
+}
+
+// exported for testing
+export function configToModels(config: KubeConfig, filePath: string): UpdateClusterModel[] {
+  const validConfigs = [];
+
+  for (const contextConfig of splitConfig(config)) {
+    const error = validateKubeConfig(contextConfig, contextConfig.currentContext);
+
+    if (error) {
+      logger.debug(`${logPrefix} context failed validation: ${error}`, { context: contextConfig.currentContext, filePath });
+    } else {
+      validConfigs.push({
+        kubeConfigPath: filePath,
+        contextName: contextConfig.currentContext,
+      });
+    }
+  }
+
+  return validConfigs;
+}
+
+type RootSourceValue = [Cluster, CatalogEntity];
+type RootSource = ObservableMap<string, RootSourceValue>;
+
+// exported for testing
+export function computeDiff(contents: string, source: RootSource, port: number, filePath: string): void {
+  runInAction(() => {
+    try {
+      const rawModels = configToModels(loadConfigFromString(contents), filePath);
+      const models = new Map(rawModels.map(m => [m.contextName, m]));
+
+      logger.debug(`${logPrefix} File now has ${models.size} entries`, { filePath });
+
+      for (const [contextName, value] of source) {
+        const model = models.get(contextName);
+
+        // remove and disconnect clusters that were removed from the config
+        if (!model) {
+          value[0].disconnect();
+          source.delete(contextName);
+          logger.debug(`${logPrefix} Removed old cluster from sync`, { filePath, contextName });
+          continue;
+        }
+
+        // TODO: For the update check we need to make sure that the config itself hasn't changed.
+        // Probably should make it so that cluster keeps a copy of the config in its memory and
+        // diff against that
+
+        // or update the model and mark it as not needed to be added
+        value[0].updateModel(model);
+        models.delete(contextName);
+        logger.debug(`${logPrefix} Updated old cluster from sync`, { filePath, contextName });
+      }
+
+      for (const [contextName, model] of models) {
+        // add new clusters to the source
+        try {
+          const cluster = new Cluster({ ...model, id: uuid.v4() });
+
+          if (!cluster.apiUrl) {
+            throw new Error("Cluster constructor failed, see above error");
+          }
+
+          cluster.init(port);
+
+          const entity = catalogEntityFromCluster(cluster);
+
+          entity.metadata.labels.KUBECONFIG_SYNC = filePath;
```

Comment:
I think we should not add this label (unless we really have to). It adds too much noise to catalog.